x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/kerberos/support/KerberosTicketValidator.java

```diff
@@ -0,0 +1,259 @@
+import java.util.Map;
+import java.util.Set;
+
+import javax.security.auth.Subject;
+import javax.security.auth.kerberos.KerberosPrincipal;
+import javax.security.auth.login.AppConfigurationEntry;
+import javax.security.auth.login.Configuration;
+import javax.security.auth.login.LoginContext;
+import javax.security.auth.login.LoginException;
+
+/**
+ * Responsible for validating Kerberos ticket<br>
+ * Performs service login using keytab, supports multiple principals in keytab.
+ */
+public class KerberosTicketValidator {
+    public static final Oid SPNEGO_OID = getSpnegoOid();
+
+    private static Oid getSpnegoOid() {
+        Oid oid = null;
+        try {
+            oid = new Oid("1.3.6.1.5.5.2");
+        } catch (GSSException gsse) {
+            ExceptionsHelper.convertToRuntime(gsse);
+        }
+        return oid;
+    }
+
+    private static final Logger LOGGER = ESLoggerFactory.getLogger(KerberosTicketValidator.class);
+
+    private static final String KEY_TAB_CONF_NAME = "KeytabConf";
+    private static final String SUN_KRB5_LOGIN_MODULE = "com.sun.security.auth.module.Krb5LoginModule";
+
+    /**
+     * Validates client kerberos ticket.
+     *
+     * @param servicePrincipalName Service principal name
+     * @param nameType {@link GSSName} type depending on GSS API principal entity
+     * @param base64Ticket base64 encoded kerberos ticket
+     * @param config {@link RealmConfig}
+     * @return {@link Tuple} of user name {@link GSSContext#getSrcName()} and out
+     *         token base64 encoded if any. When context is not yet established user
+     *         name is {@code null}.
+     * @throws LoginException thrown when service authentication fails
+     *             {@link LoginContext#login()}
+     * @throws GSSException thrown when GSS Context negotiation fails
+     *             {@link GSSException}
+     */
+    public Tuple<String, String> validateTicket(final String servicePrincipalName, Oid nameType, final String base64Ticket,
+            final RealmConfig config) throws LoginException, GSSException {
+        final GSSManager gssManager = GSSManager.getInstance();
+        GSSContext gssContext = null;
+        LoginContext loginContext = null;
+        try {
+            Path keyTabPath = config.env().configFile().resolve(KerberosRealmSettings.HTTP_SERVICE_KEYTAB_PATH.get(config.settings()));
+            if (Files.exists(keyTabPath) == false) {
+                throw new IllegalArgumentException("configured service key tab file does not exist for "
+                        + RealmSettings.getFullSettingKey(config, KerberosRealmSettings.HTTP_SERVICE_KEYTAB_PATH));
+            }
+            // do service login
+            loginContext = serviceLogin(servicePrincipalName, keyTabPath.toString(), config.settings());
+            // create credentials
+            GSSCredential serviceCreds = createCredentials(servicePrincipalName, nameType, gssManager, loginContext);
+            // create gss context
```

Comment:
I don't feel like these comments add anything. They're just repeating what is already clear from the method being called.